std/hash_map.zig

```diff
@@ -135,19 +135,17 @@ pub fn HashMap(comptime K: type, comptime V: type, comptime hash: fn (key: K) u3
             return res.kv;
         }
 
-        fn optimizedCapacity(expected_count: usize) usize {
+        fn optimizedCapacity(expected_count: usize) !usize {
             // ensure that the hash map will be at most 60% full if
             // expected_count items are put into it
             var optimized_capacity = expected_count * 5 / 3;
-            // round capacity to the next power of two
-            const pow = math.log2_int_ceil(usize, optimized_capacity);
-            return math.pow(usize, 2, pow);
+            return math.ceilPowerOfTwo(usize, optimized_capacity);
         }
 
         /// Increases capacity so that the hash map will be at most
         /// 60% full when expected_count items are put into it
         pub fn ensureCapacity(self: *Self, expected_count: usize) !void {
-            const optimized_capacity = optimizedCapacity(expected_count);
+            const optimized_capacity = try optimizedCapacity(expected_count);
```

Comment:
I believe this introduces `error.Overflow` as another possible error from the hash map put() and related functions. I think it would be better to keep the possible errors of HashMap as only `error{OutOfMemory}`. It makes sense in this case if you think about it - if such an overflow occurred the amount of memory required wouldn't even be representable in the address space.

Reply:
Done via `catch unreachable`: https://github.com/ziglang/zig/pull/2617/commits/3d5c1a550d4f8b9f75763f510586115f944183e3

Reply:
What makes the overflow unreachable? 

One idea here would be to use the ceilPowerOfTwoMinusOne. I wouldn't even be surprised if it was faster.

Reply:
What would be the difference between `catch unreachable` here and `ceilPowerOfTwoMinusOne() + 1`? Both *can* overflow and both would have essentially the same failure mode, right?

`ceilPowerOfTwoMinusOne()` will likely be marginally faster, but the `+ 1` after the call would almost completely cancel that out.

Reply:
> What would be the difference between `catch unreachable` here and `ceilPowerOfTwoMinusOne() + 1`? Both _can_ overflow and both would have essentially the same failure mode, right?

Correct, these would be semantically identical. I'm suggesting something different: to ditch the idea of `ensureCapacity` giving an actual power of two, and modifying it to round up to the nearest power of 2 minus 1. E.g. don't do the +1.

Reply:
I thought the power of two capacity was necessary for the performance of the hash map implementation? If it's irrelevant, then we could just stop at `* 5 / 3`, right?

Reply:
I just looked through the git blame history for the `ensureCapacity` function. The original logic was simply: "if we get too full (60%), double the capacity". There was never a requirement for a power of 2 - the only thing needed is to have exponential growth so that the hash map has decent amortized big O time. `capacity * 5 / 3` would be perfectly fine. Testing is needed to see what would be best, but certainly powTwoMinusOne would also be fine.

It might even be justified to do the `ceilPowerOfTwo() catch unreachable`, but I think it would at least warrant a comment explaining why the `catch unreachable` is OK.

Reply:
I've removed that change from this PR (so now it only adds the `std.math` functions) and created an issue for `ensureCapacity` power of two-ness: https://github.com/ziglang/zig/issues/2624

I'll try looking into that further.